adapter/src/main/java/com/twilio/kudu/sql/KuduEnumerable.java

```diff
@@ -495,10 +495,17 @@ public Schema getTableSchema() {
       if (!columnIndices.isEmpty()) {
         tokenBuilder.setProjectedColumnIndexes(columnIndices);
       }
-      // we can only push down the limit if we are ordering by the pk columns
-      // and if there is no offset
-      if (sort && offset == -1 && limit != -1 && !groupBySorted) {
-        tokenBuilder.limit(limit);
+      // Push down the limit if present AND
+      // 1. Not doing a group aggregation.
+      // 2. All the predicates are pushed into the scan.
+      // When those are true, the Scanners can inform the datanode that it only
+      // requires a small number of rows.
+      if (limit > 0 && !groupBySorted && filterFunction == Predicate1.TRUE) {
+        if (offset > 0) {
```

Comment:
If we have an offset we always set `sort` to true so that the rows are returned in the pk order, so we should not push down the limit in this case. 
https://github.com/twilio/calcite-kudu/blob/8a3f53bf1d292a3b3cd0be7383987a310cb67cc2/adapter/src/main/java/com/twilio/kudu/sql/KuduEnumerable.java#L134

Reply:
good catch thanks!

Reply:
I think this logic still works. 

1. Query is sorted
2. Query contains offset and limit

The `TokenBuilder#setLimit(long)` call will be `offset + limit`. Lets assume offset is 1 and limit is 2 and there are two scans each with a `.setLimit` call of `1 + 2`. 

After both Scans return, and the lowest row is from Scan A. This is discarded for the offset. The second lowest is from Scan B and the final lowest is from Scan B. Scan B  has one more record, even if the tablet that Scan B is assigned has a million more rows that would match the query, the Scan only has access to one more. Scan A has 2 more record even if their are many more records that match the query.

The scan would never receive more than `3` records from each scan. And the Union of those scans will always be greater than the `limit + offset` -- if there are enough records to satisfy the query.

Reply:
Yeah I think you would never need to return more than `limit + offset` rows from each scan.